ecosystem/lava-sdk/src/providerOptimizer/providerOptimizer.ts

```diff
@@ -0,0 +1,656 @@
+      halfTime,
+      sampleTime
+    );
+
+    if (!success && !this.providersStorage.has(providerAddress)) {
+      this.tempIgnoredProviders.add(providerAddress);
+    }
+
+    if (success && latency > 0) {
+      this.tempIgnoredProviders.delete(providerAddress);
+
+      providerData = this.updateProbeEntryLatency(
+        providerData,
+        latency,
+        this.baseWorldLatency,
+        PROBE_UPDATE_WEIGHT,
+        halfTime,
+        sampleTime
+      );
+    }
+
+    this.providersStorage.set(providerAddress, providerData);
+    Logger.debug("probe update", providerAddress, latency, success);
+  }
+
+  public appendRelayData(
+    providerAddress: string,
+    latency: number,
+    isHangingApi: boolean,
+    cu: number,
+    syncBlock: number
+  ): void {
+    this.appendRelay(
+      providerAddress,
+      latency,
+      isHangingApi,
+      true,
+      cu,
+      syncBlock,
+      now()
+    );
+  }
+
+  public appendRelayFailure(providerAddress: string): void {
+    this.appendRelay(providerAddress, 0, false, false, 0, 0, now());
+  }
+
+  private appendRelay(
+    providerAddress: string,
+    latency: number,
+    isHangingApi: boolean,
+    success: boolean,
+    cu: number,
+    syncBlock: number,
+    sampleTime: number
+  ) {
+    const { block, time } = this.updateLatestSyncData(syncBlock, sampleTime);
+    let providerData = this.getProviderData(providerAddress);
+    const halfTime = this.calculateHalfTime(providerAddress);
+
+    providerData = this.updateProbeEntryAvailability(
+      providerData,
+      success,
+      RELAY_UPDATE_WEIGHT,
+      halfTime,
+      sampleTime
+    );
+    if (success) {
+      if (latency > 0) {
+        let baseLatency = this.baseWorldLatency + baseTimePerCU(cu);
+        if (isHangingApi) {
+          baseLatency += this.averageBlockTime;
+        }
+        providerData = this.updateProbeEntryLatency(
+          providerData,
+          latency,
+          baseLatency,
+          RELAY_UPDATE_WEIGHT,
+          halfTime,
+          sampleTime
+        );
+      }
+
+      if (syncBlock > providerData.syncBlock) {
+        providerData.syncBlock = syncBlock;
+      }
+
+      const syncLag = this.calculateSyncLag(
+        block,
+        time,
+        providerData.syncBlock,
+        sampleTime
+      );
+      providerData = this.updateProbeEntrySync(
+        providerData,
+        syncLag,
+        this.averageBlockTime,
+        halfTime,
+        sampleTime
+      );
+    }
+
+    this.providersStorage.set(providerAddress, providerData);
+    this.updateRelayTime(providerAddress, sampleTime);
+    Logger.debug(
+      "relay update",
+      syncBlock,
+      cu,
+      providerAddress,
+      latency,
+      success
+    );
+  }
+
+  public chooseProvider(
+    allAddresses: string[],
+    ignoredProviders: string[],
+    cu: number,
+    requestedBlock: number,
+    perturbationPercentage: number
+  ): string[] {
+    const returnedProviders: string[] = [];
```

Comment:
in the go implementation first provider is initialized to "" and here it's not, this causes a bug